{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GDI Example dataset\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The GDI project has its own specific metadata profile based HealthDCAT-AP.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we are ready to define the dataset. We can do that as a Python dictionary.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "from getpass import getpass\n",
    "\n",
    "from rdflib import URIRef\n",
    "from rdflib.namespace import DCTERMS\n",
    "\n",
    "from sempyro.dcat import AccessRights\n",
    "from sempyro.healthdcatap import (\n",
    "    HEALTHDCATAPAgent,\n",
    "    HEALTHDCATAPDataService,\n",
    "    HEALTHDCATAPDataset,\n",
    "    HEALTHDCATAPDistribution,\n",
    "    HEALTHDCATAPKind,\n",
    "    HEALTHDCATAPHdab,\n",
    "    HEALTHDCATAPPublisher,\n",
    ")\n",
    "\n",
    "# beacon sample level server is disable by now.\n",
    "\n",
    "dataset_subject = URIRef(\"http://example.com/gdi/dataset\")\n",
    "distribution_subject_af = URIRef(\"http://example.com/distribution_1\")\n",
    "dataservice_definition_subject_af = URIRef(\"http://example.com/gdi/dataserviceaf\")\n",
    "\n",
    "env = \"test\"  # test or acc\n",
    "server_fdp = f\"https://gdi-fdp-{env}.healthdata.nl\"\n",
    "beacon_server_af = f\"https://gdi-beacon-af-{env}.healthdata.nl/api/info\"\n",
    "\n",
    "contact_point_goe = HEALTHDCATAPKind(\n",
    "    hasEmail=\"mailto:info@goe.nl\",\n",
    "    contact_page=\"https://GoE.example.nl\",\n",
    "    formatted_name=[\"GoE - NL\"],\n",
    ")\n",
    "\n",
    "creator_goe = HEALTHDCATAPAgent(\n",
    "    name=[\"Genome of Europe - NL node\"],\n",
    "    identifier=\"GOE-NL-HRI-1\",\n",
    "    mbox=[\"mailto:info@goe.nl\"],\n",
    "    homepage=\"https://GoE.example.nl\",\n",
    ")\n",
    "\n",
    "hdab_definition = HEALTHDCATAPHdab(\n",
    "    name=[\"HDAB-NL\"],\n",
    "    identifier=\"HDAB-NL\",\n",
    "    mbox=[\"mailto:info@hdab.nl\"],\n",
    "    homepage=\"https://GoE.example.nl\",\n",
    "    contact_point=contact_point_goe,\n",
    ")\n",
    "\n",
    "publisher_hdab = HEALTHDCATAPPublisher(\n",
    "    name=[\"HDAB-NL\"],\n",
    "    identifier=\"HDAB-NL\",\n",
    "    mbox=[\"mailto:info@hdab.nl\"],\n",
    "    homepage=\"https://GoE.example.nl\",\n",
    "    contact_point=contact_point_goe,\n",
    ")\n",
    "\n",
    "dataset_definition = {\n",
    "    \"access_rights\": AccessRights.public,\n",
    "    \"applicable_legislation\": [\"https://data.europa.eu/eli/reg/2025/327/oj\"],\n",
    "    \"conforms_to\": [\"http://data.gdi.eu/core/p2/ExternallyGoverned\"],\n",
    "    \"health_category\": [\n",
    "        \"http://data.gdi.eu/core/p2/HealthCategoryHumanGenetic\",\n",
    "        \"http://data.gdi.eu/core/p2/HealthCategoryHumanEpigenomic\",\n",
    "        \"http://data.gdi.eu/core/p2/HealthCategoryHumanGenomic\",\n",
    "    ],\n",
    "    \"contact_point\": [contact_point_goe],\n",
    "    \"hdab\": hdab_definition,\n",
    "    \"creator\": [creator_goe],\n",
    "    \"description\": [\n",
    "        \"The first dummy GoE examplar AF-browser dataset and on 2025-11-27 we decided that also the description \"\n",
    "        \"how we arrived at this dataset will be part of the description\"\n",
    "    ],\n",
    "    \"distribution\": [str(distribution_subject_af)],\n",
    "    \"identifier\": [\"GOE-NL-HRI-1\"],\n",
    "    \"legal_basis\": [\"https://w3id.org/dpv#Consent\"],\n",
    "    \"number_of_records\": 12340,\n",
    "    \"number_of_unique_individuals\": 1234,\n",
    "    \"publisher\": [publisher_hdab],\n",
    "    \"theme\": [\"https://publications.europa.eu/resource/authority/data-theme/HEAL\"],\n",
    "    \"type\": [\"https://publications.europa.eu/resource/authority/dataset-type/SYNTHETIC_DATA\"],\n",
    "    \"title\": [\"Genome of Europe Dutch Dummy dataset\"],\n",
    "    \"keyword\": [\"Genome of Europe\", \"GDI\"],\n",
    "}\n",
    "\n",
    "\n",
    "distribution_definition_af = {\n",
    "    \"title\": [\"Example VCF from BSC\"],\n",
    "    \"description\": [\"VCF file containing GWAS and allele frequency lookup data of synthetic Colorectal cancer cases.\"],\n",
    "    \"access_url\": [\"https://rems-goe.nl/2323\"],\n",
    "    \"media_type\": \"https://www.iana.org/assignments/media-types/application/vcf\",\n",
    "    \"license\": \"https://creativecommons.org/licenses/by-sa/4.0/\",\n",
    "    \"format\": \"https://publications.europa.eu/resource/authority/file-type/VCF\",\n",
    "    \"byte_size\": 1024,\n",
    "    \"applicable_legislation\": [\"http://data.europa.eu/eli/reg/2025/327/oj\"],\n",
    "}\n",
    "\n",
    "\n",
    "dataservice_definition_af = {\n",
    "    \"title\": [\"GoE NL Beacon\"],\n",
    "    \"description\": [\"Beacon allele frequency queries on cancer genomics data\"],\n",
    "    \"endpoint_url\": [f\"{beacon_server_af}\"],\n",
    "    \"endpoint_description\": [f\"{beacon_server_af}\"],\n",
    "    \"serves_dataset\": [str(dataset_subject)],\n",
    "    \"access_rights\": AccessRights.public,\n",
    "    \"contact_point\": [contact_point_goe],\n",
    "    \"identifier\": [\"GOE-NL-HRI-2\"],\n",
    "    \"license\": \"https://creativecommons.org/licenses/by-nc/4.0/\",\n",
    "    \"publisher\": [publisher_hdab],\n",
    "    \"theme\": [\"https://publications.europa.eu/resource/authority/data-theme/HEAL\"],\n",
    "    \"keyword\": [\"beacon v2\", \"allele frequency beacon\"],\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we instantiate the dataset class and print the serialization.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_dataset = HEALTHDCATAPDataset(**dataset_definition)\n",
    "example_dataset_graph = example_dataset.to_graph(URIRef(\"http://example.com/gdi/dataset\"))\n",
    "\n",
    "\n",
    "example_distribution_af = HEALTHDCATAPDistribution(**distribution_definition_af)\n",
    "example_distribution_graph_af = example_distribution_af.to_graph(distribution_subject_af)\n",
    "example_dataservice_af = HEALTHDCATAPDataService(**dataservice_definition_af)\n",
    "example_dataservice_graph_af = example_dataservice_af.to_graph(dataservice_definition_subject_af)\n",
    "\n",
    "# Add them up for prettier visualization\n",
    "print((example_dataset_graph + example_distribution_graph_af + example_dataservice_graph_af).serialize(format=\"turtle\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we can push the Dataset to a FAIR Data Point. For this, we use the Health-RI developed\n",
    "[FAIRClient](https://github.com/Health-RI/fairclient) library.\n",
    "\n",
    "First, we define a couple of settings. Note: you will need an existing Catalog in the FDP to add\n",
    "the dataset to. If you don't have one, you can easily create them using the web interface.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fdp_baseurl = server_fdp\n",
    "fdp_user = input(\"Enter FDP username: \")\n",
    "fdp_pass = getpass(prompt=\"Password: \")\n",
    "parent_catalog_id = input(\"Enter parent Catalog ID: \")\n",
    "fdp_parent_catalog = f\"{server_fdp}/catalog/{parent_catalog_id}\"\n",
    "\n",
    "print(f\"fdp:{fdp_baseurl}\")\n",
    "print(f\"parent catalog: {fdp_parent_catalog}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import fairclient.fdpclient\n",
    "\n",
    "# Log in to the FAIR Data Point\n",
    "fdpclient = fairclient.fdpclient.FDPClient(base_url=fdp_baseurl, username=fdp_user, password=fdp_pass)\n",
    "\n",
    "# Add a reference to the parent catalog to make the FDP happy\n",
    "example_dataset_graph.add((dataset_subject, DCTERMS.isPartOf, URIRef(fdp_parent_catalog)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_dataset = fdpclient.create_and_publish(\"dataset\", example_dataset_graph)\n",
    "print(new_dataset)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To conclude we can now add the distribution to the dataset. This is done by adding the distribution to the dataset graph and then pushing it to the FDP.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_distribution_graph_af.add((distribution_subject_af, DCTERMS.isPartOf, URIRef(f\"{new_dataset}\")))\n",
    "distribution_fdp_id_af = fdpclient.create_and_publish(resource_type=\"distribution\", metadata=example_distribution_graph_af)\n",
    "\n",
    "print(distribution_fdp_id_af)\n",
    "\n",
    "example_dataservice_graph_af.add((dataservice_definition_subject_af, DCTERMS.isPartOf, URIRef(f\"{distribution_fdp_id_af}\")))\n",
    "dataservice_fdp_id_af = fdpclient.create_and_publish(resource_type=\"dataservice\", metadata=example_dataservice_graph_af)\n",
    "\n",
    "print(dataservice_fdp_id_af)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: Manually setting the fdp_id of the distribution and dataservice into the dataset's distribution field and the distribution's access_service field is still required, as this has not been automated yet!\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
