{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GOE Example dataset\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The GDI project has its own specific metadata profile based HealthDCAT-AP.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we are ready to define the dataset. We can do that as a Python dictionary.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from getpass import getpass\n",
    "\n",
    "from rdflib import URIRef\n",
    "from rdflib.namespace import DCTERMS, RDF\n",
    "\n",
    "from sempyro.adms import Identifier\n",
    "from sempyro.dcat import AccessRights\n",
    "from sempyro.healthdcatap import (\n",
    "    HEALTHDCATAPAgent,\n",
    "    HEALTHDCATAPDataset,\n",
    "    HEALTHDCATAPDistribution,\n",
    "    HEALTHDCATAPKind,\n",
    "    HEALTHDCATAPPublisher,\n",
    ")\n",
    "\n",
    "dataset_subject = URIRef(\"https://gdi-fdp-prd.healthdata.nl/dataset/69d0d4a4-d70d-4131-ac1f-e74ad6e5fe5f\")\n",
    "distribution_subject = URIRef(\n",
    "    \"https://gdi-fdp-prd.healthdata.nl/distribution/ed723757-e030-4ec0-a905-996f77f68c36\"\n",
    ")\n",
    "\n",
    "contact_point_erasmus = HEALTHDCATAPKind(\n",
    "    hasEmail=\"mailto:j.vanrooij@erasmusmc.nl\",\n",
    "    formatted_name=[\"Jeroen van Rooij\"],\n",
    ")\n",
    "\n",
    "creator_erasmus = HEALTHDCATAPAgent(\n",
    "    name=[\"Erasmus MC\"],\n",
    "    identifier=\"https://ror.org/018906e22\",\n",
    "    mbox=[\"mailto:j.vanrooij@erasmusmc.nl\"],\n",
    "    homepage=\"https://erasmusmc.nl/\",\n",
    ")\n",
    "\n",
    "publisher_erasmus = HEALTHDCATAPPublisher(\n",
    "    name=[\"Erasmus MC\"],\n",
    "    identifier=\"https://ror.org/018906e22\",\n",
    "    mbox=[\"mailto:j.vanrooij@erasmusmc.nl\"],\n",
    "    homepage=\"https://erasmusmc.nl/\",\n",
    "    contact_point=contact_point_erasmus,\n",
    ")\n",
    "\n",
    "conforms_to_uri = \"http://data.gdi.eu/core/p2/ExternallyGoverned\"\n",
    "\n",
    "other_identifier_eva = Identifier(\n",
    "    notation=\"https://www.ebi.ac.uk/eva/?eva-study=PRJEB20726\",\n",
    "    schema_agency=\"EBI EVA\",\n",
    ")\n",
    "\n",
    "dataset_definition = {\n",
    "    \"access_rights\": AccessRights.public,\n",
    "    \"applicable_legislation\": [\"http://data.europa.eu/eli/reg/2025/327/oj\"],\n",
    "    \"conforms_to\": [conforms_to_uri],\n",
    "    \"contact_point\": [contact_point_erasmus],\n",
    "    \"creator\": [creator_erasmus],\n",
    "    \"description\": [\n",
    "        \"All variants detected by whole exome sequencing of 2628 Dutch healthy elderly individuals\"\n",
    "    ],\n",
    "    \"distribution\": [str(distribution_subject)],\n",
    "    \"identifier\": [\"GOE-NL-ERS-0001\"],\n",
    "    \"other_identifier\": [other_identifier_eva],\n",
    "    \"keyword\": [\"Cohort Study\"],\n",
    "    \"language\": [\"http://publications.europa.eu/resource/authority/language/ENG\"],\n",
    "    \"license\": \"http://purl.org/NET/rdflicense/cc-zero1.0\",\n",
    "    \"publisher\": [publisher_erasmus],\n",
    "    \"theme\": [\"http://publications.europa.eu/resource/authority/data-theme/HEAL\"],\n",
    "    \"health_category\": [\"http://data.gdi.eu/core/p2/HealthCategoryHumanGenetic\"],\n",
    "    \"title\": [\"Exome Sequencing\"],\n",
    "}\n",
    "\n",
    "# Note: HEALTHDCATAPDistribution requires http/https URLs, so the FTP URLs are\n",
    "# normalized to https for validation.\n",
    "distribution_access_url = (\n",
    "    \"https://ftp.sra.ebi.ac.uk/vol1/analysis/ERZ407/ERZ407864/170419_RSX2_EVA.EVA.vcf.gz\"\n",
    ")\n",
    "\n",
    "distribution_definition = {\n",
    "    \"title\": [\"Allele Frequencies of 2628 Exome Sequenced Samples.\"],\n",
    "    \"description\": [\n",
    "        \"All variants detected by whole exome sequencing of 2628 Dutch healthy elderly individuals\"\n",
    "    ],\n",
    "    \"access_url\": [distribution_access_url],\n",
    "    \"download_url\": [distribution_access_url],\n",
    "    \"byte_size\": 6000000,\n",
    "    \"format\": \"https://edamontology.org/format_3016\",\n",
    "    \"rights\": [\"https://www.ebi.ac.uk/eva/?eva-study=PRJEB20726\"],\n",
    "    \"license\": \"https://definities.geostandaarden.nl/DCAT-AP-NL/id/waarde/licentieValue/niet-open\",\n",
    "    \"applicable_legislation\": [\"http://data.europa.eu/eli/reg/2025/327/oj\"],\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we instantiate the dataset class and print the serialization.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_dataset = HEALTHDCATAPDataset(**dataset_definition)\n",
    "example_dataset_graph = example_dataset.to_graph(dataset_subject)\n",
    "\n",
    "# Satisfy GDI SHACL: dcterms:conformsTo must point to a dcterms:Standard\n",
    "example_dataset_graph.add((URIRef(conforms_to_uri), RDF.type, DCTERMS.Standard))\n",
    "\n",
    "example_distribution = HEALTHDCATAPDistribution(**distribution_definition)\n",
    "example_distribution_graph = example_distribution.to_graph(distribution_subject)\n",
    "\n",
    "# Add them up for prettier visualization\n",
    "print((example_dataset_graph + example_distribution_graph).serialize(format=\"turtle\"))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we can push the Dataset to a FAIR Data Point. For this, we use the Health-RI developed\n",
    "[FAIRClient](https://github.com/Health-RI/fairclient) library.\n",
    "\n",
    "First, we define a couple of settings. Note: you will need an existing Catalog in the FDP to add\n",
    "the dataset to. If you don't have one, you can easily create them using the web interface.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fdp_baseurl = input(\"Enter FDP server url: \")\n",
    "fdp_user = input(\"Enter FDP username: \")\n",
    "fdp_pass = getpass(prompt=\"Password: \")\n",
    "parent_catalog_id = input(\"Enter parent Catalog ID: \")\n",
    "fdp_parent_catalog = f\"{fdp_baseurl}/catalog/{parent_catalog_id}\"\n",
    "\n",
    "print(f\"fdp:{fdp_baseurl}\")\n",
    "print(f\"parent catalog: {fdp_parent_catalog}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import fairclient.fdpclient\n",
    "\n",
    "# Log in to the FAIR Data Point\n",
    "fdpclient = fairclient.fdpclient.FDPClient(base_url=fdp_baseurl, username=fdp_user, password=fdp_pass)\n",
    "\n",
    "# Add a reference to the parent catalog to make the FDP happy\n",
    "example_dataset_graph.add((dataset_subject, DCTERMS.isPartOf, URIRef(fdp_parent_catalog)))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_dataset = fdpclient.create_and_publish(\"dataset\", example_dataset_graph)\n",
    "print(new_dataset)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To conclude we can now add the distribution to the dataset. This is done by adding the distribution to the dataset graph and then pushing it to the FDP.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_distribution_graph.add((distribution_subject, DCTERMS.isPartOf, URIRef(f\"{new_dataset}\")))\n",
    "distribution_fdp_id = fdpclient.create_and_publish(resource_type=\"distribution\", metadata=example_distribution_graph)\n",
    "\n",
    "print(distribution_fdp_id)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: Manually setting the `fdp_id` of the distribution into the dataset's distribution field is still required, as this has not been automated yet!\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
